File: 0404-sum-of-left-leaves/0404-sum-of-left-leaves.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumOfLeftLeaves(self, root: Optional[TreeNode]) -> int:
        
        if not root: return 0
        
        def is_leaf(node):
            return node is not None and node.left is None and node.right is None
        
        stack, total = [root], 0
        
        while stack:
            sub_root = stack.pop()
            
            if is_leaf(sub_root.left):
                total += sub_root.left.val
                
            if sub_root.right:
                stack.append(sub_root.right)
            
            if sub_root.left:
                stack.append(sub_root.left)
                
        return total
        
            

File: 0404-sum-of-left-leaves/0404-sum-of-left-leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool is_leaf(TreeNode* node){
        return (node != NULL) && (node -> left == NULL) && (node -> right == NULL);
    }
    
    int sumOfLeftLeaves(TreeNode* root) {
        if (root == NULL) return 0;
        stack<TreeNode*> st({root});
        int total = 0;
        TreeNode* sub_root;
        while (!st.empty()){
            sub_root = st.top();
            st.pop();
            if (is_leaf(sub_root->left)) total += sub_root->left->val;
            
            if (sub_root->right != NULL) st.push(sub_root -> right);
            if (sub_root -> left != NULL) st.push(sub_root -> left);
        }
        
        return total;
        
    }
};